pipeline_dp/pipeline_functions.py

```diff
@@ -97,3 +97,26 @@ def create_key_fn(key):
         f"{stage_name}: list of inputs to dictionary of inputs")
     return backend.map(input_dict, lambda d: container_class(**d),
                        f"{stage_name}: construct container class from inputs")
+
+
+def max_element(backend: pipeline_backend.PipelineBackend, col,
+                stage_name: str):
+    return choose_element(backend, col, max, stage_name)
+
+
+def min_element(backend: pipeline_backend.PipelineBackend, col,
+                stage_name: str):
+    return choose_element(backend, col, min, stage_name)
+
+
+T = TypeVar('T')
+
+
+def choose_element(backend: pipeline_backend.PipelineBackend, col,
+                   choosing_function: Callable[[T, T], T], stage_name: str):
+    return backend.values(
```

Comment:
here, I have question about efficiency and scalability. My first implementation was `backend.map(backend.to_list(col), lambda l: max(l))` but this is obviously bad and won't fit into the memory.

This implementation seems okaish and will work on Beam (https://beam.apache.org/documentation/transforms/java/aggregation/combine/) but there is also a dedicated way to get max and min (https://beam.apache.org/documentation/transforms/python/aggregation/max/) by using Combine.Globally(). I wonder whether we should implement it framework agnostic and rely on `reduce_per_key` implementation or define max/min functions in backend interface and implement them separately for each framework. WDYT?

Reply:
That's a good question. I agree that it makes sense to compute with `reduce_per_key`.
What do you think about computing both min and max at one step (which should be better from performance point of view), namely

```
# let col has elements floats
col = backend.map(col, lambda x: (None,(x, x))) # None is dummy key
col = backend.reduce_per_key(col, lambda x, y: (min(x[0], y[0]), min(x[1], y[1])) 
```
then no need to flatten and combine min and max later.

WDYT?

Reply:
hm, this approach is definitely the most efficient one we can do, however I wonder whether Beam will be smart and fuse these 2 operation, so each element of the collection will be processed only once and min/max functions will be applied with one read. The advantage is that we have min and max separately and it makes interface is cleaner.

However, since this is the first case when we need min and max and also most probably in our cases we will need always both, I guess we should save time checking whether beam fuses these 2 operations or not and just implement it as you proposed. Will do it.